grug/testing/tests/indexing.rs

```diff
@@ -273,3 +280,235 @@ fn no_sql_index_error_after_restart() {
+        Fail { err: String },
+        /// Insert the given string into storage; then, call self with the given
+        /// execute message.
+        Perform {
+            deep: String,
+            // Must be boxed due to being a recursive type.
+            next: Box<ExecuteMsg>,
+            reply_on: ReplyOn,
+        },
+    }
+
+    impl ExecuteMsg {
+        pub fn ok<T>(deep: T) -> Self
+        where
+            T: Into<String>,
+        {
+            Self::Ok { deep: deep.into() }
+        }
+
+        pub fn perform<T>(deep: T, next: ExecuteMsg, reply_on: ReplyOn) -> Self
+        where
+            T: Into<String>,
+        {
+            Self::Perform {
+                deep: deep.into(),
+                next: Box::new(next),
+                reply_on,
+            }
+        }
+    }
+
+    #[derive(Serialize, Deserialize)]
+    pub enum QueryMsg {
+        Data {},
+    }
+
+    pub struct QueryDataRequest {}
+
+    impl QueryRequest for QueryDataRequest {
+        type Message = QueryMsg;
+        type Response = Vec<String>;
+    }
+
+    impl From<QueryDataRequest> for QueryMsg {
+        fn from(_req: QueryDataRequest) -> Self {
+            Self::Data {}
+        }
+    }
+
+    pub const DEEPTHS: Set<String> = Set::new("s");
+
+    pub fn instantiate(_ctx: MutableCtx, _msg: Empty) -> StdResult<Response> {
+        Ok(Response::new())
+    }
+
+    pub fn execute(ctx: MutableCtx, msg: ExecuteMsg) -> StdResult<Response> {
+        match msg {
+            ExecuteMsg::Fail { err } => {
+                // We don't have a generic error as in CosmWasm, so use host
+                // error to mock it.
+                Err(StdError::host(err))
+            },
+            ExecuteMsg::Ok { deep } => {
+                DEEPTHS.insert(ctx.storage, deep)?;
+
+                Ok(Response::new())
+            },
+            ExecuteMsg::Perform {
+                deep,
+                next,
+                reply_on,
+            } => {
+                DEEPTHS.insert(ctx.storage, deep)?;
+
+                Ok(Response::new().add_submessage(SubMessage {
+                    msg: Message::execute(ctx.contract, &*next, Coins::new())?,
+                    reply_on,
+                }))
+            },
+        }
+    }
+
+    pub fn query(ctx: ImmutableCtx, msg: QueryMsg) -> StdResult<Json> {
+        match msg {
+            QueryMsg::Data {} => {
+                let res = DEEPTHS
+                    .range(ctx.storage, None, None, Order::Ascending)
+                    .collect::<StdResult<Vec<_>>>()?;
+                res.to_json_value()
+            },
+        }
+    }
+
+    pub fn reply(ctx: SudoCtx, msg: ReplyMsg, res: SubMsgResult) -> StdResult<Response> {
+        let msg = match (res, msg) {
+            (GenericResult::Err(_), ReplyMsg::Fail(execute_msg)) => execute_msg,
+            (GenericResult::Ok(_), ReplyMsg::Ok(execute_msg)) => execute_msg,
+            _ => panic!("invalid reply"),
+        };
```

Comment:
The matching arms can be combined:

```suggestion
        let msg = match (res, msg) {
            (GenericResult::Err(_), ReplyMsg::Fail(execute_msg))
            | (GenericResult::Ok(_), ReplyMsg::Ok(execute_msg)) => execute_msg,
            _ => panic!("invalid reply"),
        };
```